modules/highlight/highlight.go

```diff
@@ -179,61 +177,90 @@ func File(numLines int, fileName, language string, code []byte) []string {
 
 	iterator, err := lexer.Tokenise(nil, string(code))
 	if err != nil {
-		log.Error("Can't tokenize code: %v", err)
-		return plainText(string(code), numLines)
+		return nil, fmt.Errorf("can't tokenize code: %w", err)
 	}
 
-	err = formatter.Format(htmlw, styles.GitHub, iterator)
+	err = formatter.Format(htmlWriter, styles.GitHub, iterator)
 	if err != nil {
-		log.Error("Can't format code: %v", err)
-		return plainText(string(code), numLines)
-	}
-
-	htmlw.Flush()
-	finalNewLine := false
-	if len(code) > 0 {
-		finalNewLine = code[len(code)-1] == '\n'
+		return nil, fmt.Errorf("can't format code: %w", err)
 	}
 
-	m := make([]string, 0, numLines)
-	for i, v := range strings.SplitN(htmlbuf.String(), "\n", numLines) {
-		content := string(v)
+	_ = htmlWriter.Flush()
 
-		// remove useless wrapper nodes that are always present
-		content = strings.Replace(content, "<span class=\"line\"><span class=\"cl\">", "", 1)
-		content = strings.TrimPrefix(content, `</span></span>`)
+	m := make([]string, 0, bytes.Count(code, []byte{'\n'})+1)
 
-		// if there's no final newline, closing tags will be on last line
-		if !finalNewLine && i == numLines-1 {
-			content = strings.TrimSuffix(content, `</span></span>`)
+	htmlStr := htmlBuf.String()
+	line := strings.Builder{}
+	insideLine := 0 // every <span class="cl"> makes it increase one level, every closed <span class="cl"> makes it decrease one level
+	tagStack := make([]string, 0, 4)
+	for len(htmlStr) > 0 {
+		pos1 := strings.IndexByte(htmlStr, '<')
+		pos2 := strings.IndexByte(htmlStr, '>')
+		if pos1 == -1 || pos2 == -1 || pos1 > pos2 {
+			break
 		}
-
-		// need to keep lines that are only \n so copy/paste works properly in browser
-		if content == "" {
-			content = "\n"
-		} else if content == `</span><span class="w">` {
-			content += "\n"
+		tag := htmlStr[pos1 : pos2+1]
+		if insideLine > 0 {
+			line.WriteString(htmlStr[:pos1])
 		}
-
-		m = append(m, content)
+		if tag[1] == '/' {
+			if len(tagStack) == 0 {
+				return nil, fmt.Errorf("can't find matched tag: %q", tag)
+			}
+			popped := tagStack[len(tagStack)-1]
+			tagStack = tagStack[:len(tagStack)-1]
+			if popped == `<span class="cl">` {
+				insideLine--
+				lineStr := line.String()
+				if lineStr != "" && lineStr[len(lineStr)-1] == '\n' {
+					lineStr = lineStr[:len(lineStr)-1] + "&#10;"
```

Comment:
Newline needs no escaping in HTML, we can just put literal `\n`.

Reply:
This (and below) is by purpose, to make developers know that this new line is necessary for rendering, to distinguish from other ignorable/trim-able new lines.

Then it's easier to read the output, to find the unexpected newlines and to write test code for it. It won't waste too much bandwidth, because normal code and highlight tags are already very (much more) large, the only one `&#10;` for each line won't bring negative effects.

Reply:
I do hope this replacement doesn't cause incompatibilties with frontend code that might specifically look for literal `\n` instead of `&#10;`. I don't think our first-party code has any of that, but 3rd party code might. 

Should definitely test mermaid renderer after this change for example.

Reply:
That's HTML. In HTML, `&#10;` is `\n`, there should be no different for any parser.

For mermaid, the code is :

```js
const source = el.textContent;
valid = mermaid.parse(source);
```

It uses the correctly unescaped original text for source.

If there is any bug, we can still handle it easily, either do a replacement on our side, or just change the `&#10;` back to `\n`, it should be pretty easy.

Reply:
Ah, mermaid renderer actually shouldn't be affected as it takes to source out of rendered markdown, which isn't changed here.